### ISSUE TITLE
Adding forecasting backtest/evaluate function 

### ISSUE THREAD

**Issue by aiwalter** [alan-turing-institute/sktime#622]
**Is your feature request related to a problem? Please describe.**
I would like to add a backtesting class. We should ideally then also implement an expanding window splitter functionality as discussed in #552 . But the `Backtest` would just receive a splitter as an argument.

**Describe the solution you'd like**
A class like `Backtest` (naming to be discussed) with a signature like this:
```python
class Backtest(object):
    """Backtest forecasters

    Parameters
    ----------
    forecaster: estimator object
    splitter: splitter object
        e.g. SlidingWindowSplitter
    score: function
        A function like smape_loss that has y_pred, y_test as arguments and returns a float

    Returns
    ----------
    error: float
        Average error of fitted estimators calculated by means of the score function
    """
```

The class should have a `fit()` function like a regular estimator object. I guess a `predict()` function does not make sense here. The class will be able to also take a `forecaster` of type `ForecastingGridSearchCV`, as in production people might refit and tune the model by progress in time.

Optionally, the class can print the number of fits, like in `ForecastingGridSearchCV` when fitting.

I am not sure on how to structure the return and what else should be returned, I guess just return `self` and have the rest as class variables? We could do sth like in `ForecastingGridSearchCV` with the `cv_results_` variable, so we could also store there the errors, predictions  and also the params of each iteration (because params might differ when using a grid search as forecaster).

![image](https://user-images.githubusercontent.com/29627036/104709492-4ea57100-571f-11eb-9311-bf3920db1bcf.png)

Where should a class `Backtest` be placed in the `sktime` class structure? I propose `sktime/sktime/forecasting/model_backtest/_backtest.py`.

**Comment by aiwalter** [comment]
we could actually also have a `predict()` function that just takes the last fitted estimator and calls its `predict()` then. This might be useful in productive scenarios, so when you deliver a productive model you can just use a `Backtest` forecaster and deliver the backtested error together with the next prediction. So you dont have to do both separately.

**Comment by mloning** [comment]
@aiwalter I'm not familiar with this type of class, could you describe its purpose a bit more, what's the use case, what's the user journey/workflow you have in mind?

**Comment by ngupta23** [comment]
@mloning Here is [one example](https://unit8co.github.io/darts/examples/darts-intro.html#Backtesting:-simulate-historical-forecasting), but I am not sure how this is different from performing validation using [`SlidingWindowSplitter`](https://www.sktime.org/en/latest/api_reference/modules/auto_generated/sktime.forecasting.model_selection.SlidingWindowSplitter.html#sktime.forecasting.model_selection.SlidingWindowSplitter) in `sktime`.

I think one difference may be that in backtesting, you would refit the model in at every split which is currently not implemented in `sktime` per https://github.com/alan-turing-institute/sktime/issues/621#issuecomment-760228618. I would also be interested in hearing more thoughts from @aiwalter on the backtesting topic.

**Comment by aiwalter** [comment]
@mloning @ngupta23 
Actually the `Backtest` class would be similar to `GridSearchCV`. However I tried to use grid search to do a backtest and didnt get the expected result. Lets start with what I want to do and the worklflow of `Backtest`, then below I will give the example I tried with `GridSearchCV`.

1) Assumptions
Lets do some assumptions for an example to make it cleear:
- We take the airline data from the notebook example (`load_airline()`) with `freq=M`
- I want to build a model with `fh=[1,2,3,4,5,6,7,8,9,10,11,12]`, so basically forecast all months of the next year on the end of each year.

2) Workflow:
With this assumptions I want to know how well my trained model is performing, so I can do a train-test-split and take the last 12 months as test data. However, this can tell me only how the model would have performed in the last year. If I take now a bigger test data with last 24 months, I violate my assumption of only forecast 12 month (and the additional 12 months will have also a worse error, so this will be a bias). Here comes the need for a `Backtest` class (and also an expanding window splitter), because e.g. I want to backtest my model on last 5 years. So here a sketch on this example, with an expanding window splitter:

![image](https://user-images.githubusercontent.com/29627036/104759607-71a44500-5760-11eb-9dc4-76f2c517db5e.png)

Within this 5 model fits, I also want also to be able to do a grid search. This is because in production I also would most likely do it after each year because I have new data and might find better hyperparameters. 

3) Example using `GridSearchCV` to backtest:
Now what I tried so far. 
```python
cv = SlidingWindowSplitter(
    initial_window=84,
    window_length=12,
    step_length=12)

gscv = ForecastingGridSearchCV(
    forecaster=TBATS(), param_grid={"sp": [12]}, cv=cv, scoring=sMAPE()
)
gscv.fit(y_train)
pd.DataFrame(gscv.cv_results_)
```
![image](https://user-images.githubusercontent.com/29627036/104764707-0f4f4280-5768-11eb-9238-3a57d5b41cca.png)

In case I dont want to do grid search but just use the `SlidingWindowSplitter`, I have to anyway provide `param_grid={"sp": [12]}` which looks a bit strange, because I would like to just do `TBATS(sp=12)` without providing `param_grid` values. I figured out I can just do `param_grid={}` to get it running. The main point is however rather the `cv_results_`, where I would like to see 5 rows now because I have fitted 5 models and I need to exactly know how each of them has performed. Further I would like to get the average error for the 5 years, this is what is currently returned in this 1 row I think. The `Backtest` class could then also have plot function to make a nice plot of the backtest.

Btw can s.o. explain me the `initial_window` argument of `SlidingWindowSplitter` in context of my example? 😃  I am not sure if I have set this value correctly here? 

In case I want to do grid search, I would have to nest two grid searches:
```python
cv_1 = SlidingWindowSplitter(
    initial_window=84,
    window_length=12,
    step_length=1)

cv_2 = SlidingWindowSplitter(
    initial_window=84,
    window_length=12,
    step_length=12)

gscv_1 = ForecastingGridSearchCV(
    forecaster=TBATS(), param_grid={"sp": [6, 12], "use_box_cox": [True, False]}, cv=cv, scoring=sMAPE()
)

gscv_2 = ForecastingGridSearchCV(
    forecaster=gscv_1, param_grid={}, cv=cv, scoring=sMAPE()
)
gscv.fit(y_train)
pd.DataFrame(gscv.cv_results_)
```
![image](https://user-images.githubusercontent.com/29627036/104764776-242bd600-5768-11eb-8195-9dafefcef548.png)

Btw I am wondering why there is no `use_box_cox` in the params column?

So long story short: From UX and what I would like to have as return it would be nice to have a `Backtest` class. Backtesting with `ForecastingGridSearchCV` is not the best idea I think. Also I think what @fkiraly said in #552 makes sense to me:

>  we should be careful and not mix evaluation and tuning. Tuning are wrappers/compositors like GridSearchCV in sklearn; evaluation should be external to any estimator.

Happy to get your feedback.

**Comment by ngupta23** [comment]
@aiwalter , Here are my initial thoughts and understanding on this. 

I think your use of `initial_window` is more or less Ok (you may want to use 72 instead of 84 though because of the following) - 
* For each of the candidate parameters in the `param_grid`, the model is trained on the `initial_window` first. 
* Then the `window_length` of data is appended to the `initial_window` data (essentially expanding window, if you use `start_with_window`=`True`) and the prediction is made which is used for evaluation. So in your example, data will be expanded from 72 to 84 and then future prediction will be made beyond that and those will be used for evaluation. 
* This process is repeated until you reach the end of the data (keep appending window length of data and making prediction and evaluation.

There are a few caveats though based on my analysis. 

* I found that in some (maybe all?) cases, it is only doing 1 step ahead prediction in the expanding window phase. When I tried adding a forecast horizon of 12 points, it gave an error that this is not supported yet.
* When the window is expanded, in some cases (maybe all?), the model parameters are not refit. It retains the same model parameters that were learned in the `initial_window`. See discussion on this topic here (https://github.com/alan-turing-institute/sktime/issues/621). I think for backtesting, you want to refit it at each expanding window.

Also, regarding the following...

> Btw I am wondering why there is no use_box_cox in the params column?

The params column will only show the params that you have provided in the grid search. Default/implicit parameters are not shown in `cv_results_`.

**Comment by fkiraly** [comment]
@aiwalter, thanks for your suggestion - I think having robust functionality for quick evaluation and backtesting is a great idea.

However, in summary of my answer below:
* I do *not* think backtesting should be a class like `GridSearchCV`, butpart of  an "external" benchmarking module
* it would be great if you would like to contribute to building that!

So, why should backtesting not be a class like `GridSearchCV`?
Because backtesting, or more generally performance evaluation workflows, are not the same as tuning strategies.

A tuning strategy still is a forecaster/estimator, but a performance evaluation workflow isn't. A forecaster is subjected to a performance evaluation workflow. Both may use re-sampling (e.g., cross-validation or temporal re-sampling), but still they aren't the same kinds of formal objects.

For the same reason, `sklearn`'s `GridSearchCV` should not be used for benchmarking or performance evaluation, because instances of it are representations of estimators, not evaluation workflows.

Some more detail on this, in my opinion, crucial distinction between tuning and evaluation workflows can be found in Section 5.1 of this paper on AI toolbox software design: https://arxiv.org/abs/2101.04938 (compare: anti-pattern 4; we seem to have been pre-scient here :-) )

Somewhat confusingly, the same re-sampling technique can be used in a tuning workflow as well as in an evaluation workflow.
For example, cross-validation is used inside `GridSearchCV` - to obtain performance estimates which are then used in tuning - and in evaluation workflows - to obtain performance estimates.

In terms of software design, the former should be a class (with a higher-order scitype), but hte latter shouldn't be, even if both use the same re-sampling technique.
Evaluation should be on the workflow layer, and calls subject to something like
`benchmark(estimator, Backtesting(config), metrics)`
or
`Benchmarking.run()` (after construction with config setting)

Does this distinction make sense?

Btw, @ViktorKaz has been working on the design of benchmarking workflows for a while.
The newest for `sktime` is in this extension proposal here: https://github.com/sktime/enhancement-proposals/pull/5
I would suggest you think of contributing and/or designing the evaluation workflow(s)?

There is an also earlier toolbox we've written for benchmarking of `sklearn` vanilla estimators, but it's still quite immature:
https://arxiv.org/abs/1901.03678
The idea was to re-design, and also introduce an abstract benchmarking interface for `sktime` at the same time.

**Comment by koralturkk** [comment]
@aiwalter 

I believe what you would like to achieve can be simply implemented by extending adding some features to ForecastingGridSearchCV class and by creating an ExpandingWindowSplitter class. 

`cv = ExpandingWindowSplitter(
    initial_window=84,
    window_length=12)

gscv = ForecastingGridSearchCV(
    forecaster=TBATS(), param_grid={"sp": [12]}, cv=cv, scoring=sMAPE()
)
gscv.fit(y_train)
pd.DataFrame(gscv.cv_results_)`

Ideally this instantiation will have a growing training set performing fitting on a dynamic training set and logging errors for each iteration. Then we may store the results and errors of each iteration in "cv_results_" as well as the mean which would provide an extensive view to the user of how each model behaves at different stages. Satisfying the proposed workflow that you have suggested. Additionally, we may extend how logging takes place in sliding window as well to capture also the behaviors as well. This would enable us to figure out whether there are any anomalies when the training set is changed and a new forecast has been made which maybe hidden from the user in the current setting just because the mean value is satisfactory.

**Comment by aiwalter** [comment]
@fkiraly thanks for your comment here. Before I start coding, one more question: What is the intention of an additional  `Benchmarking` class? I think what we need as input for a backtest is a splitter (e.g. `SlidingWindowSplitter`), a forecaster, a score function (e.g. `smape_loss`) and the data. So would it be ok to design the interface as I wrote initially?

> ```
> class Backtest(object):
>     """Backtest forecasters
> 
>     Parameters
>     ----------
>     forecaster: estimator object
>     splitter: splitter object
>         e.g. SlidingWindowSplitter
>     score: function
>         A function like smape_loss that has y_pred, y_test as arguments and returns a float
> ```

The data could be given then in `backtest.run(y)`? Or how would this look like with an additional `Benchmarking` class?

**Comment by mloning** [comment]
@aiwalter There is also the `benchmarking` module and an example notebook. We separate orchestration (fitting and predicting) from evaluation (metrics, uncertainty estimates, etc). But it currently only works for classification, not forecasting.

**Comment by aiwalter** [comment]
@mloning ok so in my idea, I planned to also fit/predict on estimators (or grid searches) inside the `Backtest` class to simulate (backtest) the regular workflow. If this is not the intention of your evaluation approach, I might better keep this class private?

**Comment by fkiraly** [comment]
@aiwalter, @mloning, since the discussion on evaluation/benchmarking vs tuning is quite subtle - how about a phone call plus anyone involved in implementation of tuning compositors, i.e., @ngupta and @pabworks?

Source of the subtleties is that evaluation is done as part of many tuning meta-algorithms.

One possible point of abstraction here is an "evaluation spec" and a "benchmarker"; the "benchmarker" can be called with an "evaluation spec" to compare algorithms; and an abstract tuning algorithm would take an "evaluation spec" as a hyper-parameter, and internally call the "benchmarker".

The "evaluation spec" may consist of a re-sampling descriptor, a performance metric (or multiple), and some instructions on aggregation over folds.

**Comment by mloning** [comment]
@aiwalter That's the same idea, the `orchestrator` takes in estimators (be it primitive algorithms or composites) and some CV specification and writes out prediction for each CV fold. The `evaluator` reads in the predictions and actual test data and computes scores, uncertainty estimates, statistical tests, etc. 

 @fkiraly Happy to have a call

**Comment by fkiraly** [comment]
aim of the call would be to agree on interface designs for benchmarking and tuning, as well as to discuss how to "pin down" the mathematical issues here, e.g., formal content documentation of "what".

**Comment by aiwalter** [comment]
@fkiraly  @mloning just let me know about the call and I will try to join

**Comment by ViktorKaz** [comment]
happy to have a call to discuss this.

**Comment by pabworks** [comment]
Also happy to jump on a call for this discussion.

**Comment by ngupta23** [comment]
@fkiraly , I will be happy to get on a call to discuss this as well.

**Comment by fkiraly** [comment]
find a time on slack? @mloning, @pabworks, @ngupta23, @aiwalter, @viktorkaz

**Comment by pabworks** [comment]
I do not believe I am part of this Slack channel - @mloning, I have sent you my personal email on Gitter. If you could please add me to the Slack channel, that would be much appreciated!

**Comment by mloning** [comment]
Created a discussion on slack to find a time that works for everyone - let me know if you're not on slack or send me a private message on Gitter.

**Comment by fkiraly** [comment]
So, to recapitulate what I've understood how we've resolved this in https://github.com/sktime/enhancement-proposals/pull/8:
* there should be a standalone function "evaluate" or similar for "quick and dirty" evaluation - this includes backtesting as a special case. Exact signature tbd, @aiwalter can make a proposal.
* separately from that, there will be advanced benchmarking/inspection functionality in line with the orchestration module, see #64
* tuning wrapper compositors - which are not the topic of this issue - will use this evaluation function internally.

From STEP01, evaluation could look like this as per @mloning's proposal:
```python
def evaluate(forecaster, y, fh, X=None, cv=None, strategy="refit", scoring=None):
```

strategy should have at least `"refit"` and `"update"` available.

**Comment by aiwalter** [comment]
@fkiraly this looks fine to me. Regarding backtesting, this would work with `evaluate()`. Important would be what this function is returning. As it is not a class, it would probably have to return a `pd.DataFrame` directly instead of `gscv.cv_results_` as in `ForecastingGridSearchCV`. 

I would propose at least the following columns of the `DataFrame`:
- All columns of  `gscv.cv_results_` (maybe not the rank one)
- `cutoff `(of the window)
- start of the window (equivalent to `cutoff` but the first index instead of last)
- `y_train`
- `y_pred`

To save the `y_train` and `y_pred` is important to create plots of the evaluation (backtesting).

Any other ideas? I could work on this then @mloning ? Where do we want to locate `evalute`, in `sktime.performance_metrics.forecasting.evaluate` ?

**Comment by fkiraly** [comment]
Hm, @aiwalter, that doesn´t quite seem to fit when I think through this, since you may nave more predictions `y_train` than `cutoff`s, if you make multi-step-ahead predictions? So the different columns would have different lengths?

It also sounds like you actually want a luxury version since you want to create plots.
How about a compromise - class but with default behaviour that gives you the "shorthand" behaviour?

```python
class BenchmarkForecast(object):
    """Backtest forecasters

    Parameters
    ----------
    forecaster: single forecaster, or list of forecasters
    splitter: splitter object to specify re-sampling
        e.g. SlidingWindowSplitter
    fh: forecasting horizon, relative; used in prediction
    benchmark_schema: string, "refit" or "update"; specifies walk-forward behaviour
    scoring: sklearn scorer object
        A function like smape_loss that has y_pred, y_test as arguments and returns a float
    averaging_fh: string or function, default=mean; specifies how to average over the forecasting horizon
    averaging_cutoff: string or function, default=mean; specifies how to average over the different cutoffs

    Methods:

    run(y, X=None)
        runs the experiment and stores intermediate results to self, e.g., sequence of losses or y_train, etc

    score(y, X=None)
        internally calls run and returns one number per forecaster (aggregate error)

    luxury_evaluate(y,X) returns cv_results, y_train etc; alternatively, get_predictions() or similar (?)
        (need to think about the name, behaviour and what is returned)
    """
```

**Comment by aiwalter** [comment]
@fkiraly at Mercedes we say "The best or nothing", and its all about luxury 😄 . And `sktime` fits to this slogan as well. 
I would appreciate going for a class. We are more flexible in what we return, as we can store stuff just in `self ` and the user can pick what he wants to.

**Comment by mloning** [comment]
I would start with a stand-alone function and simply return multiple output objects. 

Once this is working we can integrate into a more extensive benchmarking module. For this, I'd suggest we go with the current design in the benchmarking module, separating concerns of (i) model fitting and prediction (orchestration) and (ii) analysing/evaluating/comparing results (evaluation).

**Comment by fkiraly** [comment]
Alright, @mloning, after some thinking I agree with you: 

let's go with the standalone function first, no bells and whistles, no luxury functionality.
In a second step we work on the benchmarking module, which has nice plotting, hypothesis testing, etc.

Otherwise we risk creating a halfway monstrosity, falling into a "feature proliferation" trap too early.

What do you think, @aiwalter?
After all, Rome wasn't built on one day, and even Daimler built the 35 PS before designing the Maybach...

**Comment by aiwalter** [comment]
Ok, lets go then for now for the function as I am also a fan of keeping things simple. The plot is actually a nice to have, and might be just added to a notebook and not be part of the  `sktime` packe. Anyway, I might try to add an optional return argument to return the data that is required for the plot.

**Comment by fkiraly** [comment]
>Anyway, I might try to add an optional return argument to return the data that is required for the plot.

I've thought about that, but the problem that poses is that it is not intuitive that the return is a tuple if the user doesn't need the optional return. A compromise might be sth like `return_predictions` and only if this is `True` the return is multiple.

**Comment by aiwalter** [comment]
yes, default would not return a tuple. Same as with `predict()` and `predict(return_pred_int=True)`

**Comment by fkiraly** [comment]
makes sense to me.

It would be great if you could focus on implementing the core functionality though (e.g., make partial PR without the extra functionality), since we decided to make implementation of grid & random CV a conditionality.
